### PR TITLE
net: added new dynamic packet buffer implementation

### PR DESCRIPTION
**Pull request by haukepetersen**
During the gnrc network stack debugging, the existing implementation tended to lead to hard/segmentation faults (although these might not necessarily have causes inside the pktbuf code). To be on the very safe side for the packet buffer, I implemented a new one, using dynamic memory allocation. The implementation is very simple and straight forward (~200 lines vs ~500 for the existing dynamic implementation), and passes all unit tests. Also it is very simple to review. My intend is to use this implementation to ship around eventual pktbuf bugs so I can focus on on other ones...

I don't mind if we merge this implementation into master (having more then one possible implementation to chose from was always the plan), but I won't cry if we don't merge it. But what I dont want is a time consuming discussion on this... We have more pressing issues at the moment...

### CODE REVIEW

File: Makefile.dep [hunk 1]
```diff
@@ -316,3 +316,9 @@ ifneq (,$(filter cpp11-compat,$(USEMODULE)))
   USEMODULE += timex
   FEATURES_REQUIRED += cpp
 endif
+
+ifneq (,$(filter ng_pktbuf, $(USEMODULE)))
+  ifeq (,$(filter ng_pktbuf_dynamic, $(USEMODULE)))
```
Comment (Kijewski) [inline Makefile.dep]: No need for the inner guard

File: cpu/native/include/cpu_conf.h [hunk 2]
```diff
@@ -46,8 +46,8 @@ extern "C" {
 #else /* Linux etc. */
 #define THREAD_STACKSIZE_DEFAULT            (8192)
 #define THREAD_STACKSIZE_IDLE               (8192)
-#define THREAD_EXTRA_STACKSIZE_PRINTF       (8192)
-#define THREAD_EXTRA_STACKSIZE_PRINTF_FLOAT (8192)
+#define THREAD_EXTRA_STACKSIZE_PRINTF       (8112)
+#define THREAD_EXTRA_STACKSIZE_PRINTF_FLOAT (8112)
```
Comment (Kijewski) [inline cpu/native/include/cpu_conf.h]: 8112 is a rather odd power of 2 :)

File: sys/net/crosslayer/ng_pktbuf_dynamic/ng_pktbuf_dynamic.c [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+            --pkt->users;
+        }
+        pkt = tmp;
+    }
+    mutex_unlock(&lock);
+}
+
+ng_pktsnip_t *ng_pktbuf_start_write(ng_pktsnip_t *pkt)
+{
+    if (pkt) {
```
Comment (Kijewski) [inline sys/net/crosslayer/ng_pktbuf_dynamic/ng_pktbuf_dynamic.c]: Isn't `pkt == NULL` a user error?

File: cpu/native/include/cpu_conf.h (hunk 2 above)
Comment (haukepetersen) [inline cpu/native/include/cpu_conf.h]: This was a short hack - see #3167 for a better try on fixing this

File: sys/net/crosslayer/ng_pktbuf_dynamic/ng_pktbuf_dynamic.c (hunk 3 above)
Comment (haukepetersen) [inline sys/net/crosslayer/ng_pktbuf_dynamic/ng_pktbuf_dynamic.c]: yes. But as stated above, for testing purposes I tried to make this implementation as error prone as possible... Some of the mutex locks might also no be needed, but I put them in regardless...

File: Makefile.dep (hunk 1 above)
Comment (haukepetersen) [inline Makefile.dep]: Yes, there is, right? If I would take out the inner guard, `ng_pktbuf_default` would be included everytime `ng_pktbuf` is defined. `USEMODULE+=ng_pktbuf_dynamic` would then have no effect...

File: sys/net/crosslayer/ng_pktbuf_dynamic/ng_pktbuf_dynamic.c [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+                         ng_nettype_t type, ng_pktsnip_t *pkt)
+{
+    ng_pktsnip_t *snip;
+    void *tmp;
+
+    snip = (ng_pktsnip_t *)malloc(sizeof(ng_pktsnip_t));
+    tmp = malloc(size);
+    if ((snip == NULL) || (tmp == NULL)) {
+        free(snip);
+        free(tmp);
```
Comment (authmillenon) [inline sys/net/crosslayer/ng_pktbuf_dynamic/ng_pktbuf_dynamic.c]: Here and below: I'm not sure if free'ing NULL pointers works.

File: sys/net/crosslayer/ng_pktbuf_dynamic/ng_pktbuf_dynamic.c [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+/* for testing */
+#ifdef TEST_SUITES
+bool ng_pktbuf_is_empty(void)
+{
+    return used_sum == 0;
+}
+
+void ng_pktbuf_reset(void) {
+    count = 0;
+    used_sum = 0;
```
Comment (authmillenon) [inline sys/net/crosslayer/ng_pktbuf_dynamic/ng_pktbuf_dynamic.c]: This doesn't really reset the packet buffer ;-)

File: sys/Makefile [hunk 6]
```diff
@@ -170,6 +167,12 @@ endif
 ifneq (,$(filter ng_netdev_eth,$(USEMODULE)))
     DIRS += net/link_layer/ng_netdev_eth
 endif
+ifneq (,$(filter ng_pktbuf_default,$(USEMODULE)))
+    DIRS += net/crosslayer/ng_pktbuf_default
+endif
+ifneq (,$(filter ng_pktbuf_dynamic,$(USEMODULE)))
+    DIRS += net/crosslayer/ng_pktbuf_dynamic
+endif
```
Comment (authmillenon) [inline sys/Makefile]: I don't understand your system for sys/Makefile... This however is prone to conflict with other PRs that just squish the path defintitions to the end

**Comment by authmillenon** [comment]
Hauke: I'd like to take your implementation (since I also plan to simplify the default packet buffer), but you have to clean this PR up a little. Specifically:

- [ ] take out unrelated change to `cpu_conf.h`
- [ ] add seperate unittests for `pktbuf_dynamic`.

**Comment by authmillenon** [comment]
Oh… and default unittests are failing on my system btw:

```
$ USEMODULE="ng_pktbuf_dynamic" make -C tests/unittests/ tests-pktbuf test
[…]
..
ng_pktbuf_tests.test_pktbuf_add__pkt_NULL__data_NULL__size_0 (/home/martine/Repositories/RIOT-OS/RIOT/tests/unittests/tests-pktbuf/tests-pktbuf.c 63) pkt->data was not null.
.
ng_pktbuf_tests.test_pktbuf_add__pkt_NOT_NULL__data_NULL__size_0 (/home/martine/Repositories/RIOT-OS/RIOT/tests/unittests/tests-pktbuf/tests-pktbuf.c 80) pkt->data was not null.
.
ng_pktbuf_tests.test_pktbuf_add__pkt_NULL__data_NOT_NULL__size_0 (/home/martine/Repositories/RIOT-OS/RIOT/tests/unittests/tests-pktbuf/tests-pktbuf.c 103) !ng_pktbuf_is_empty()
.....
ng_pktbuf_tests.test_pktbuf_add__in_place (/home/martine/Repositories/RIOT-OS/RIOT/tests/unittests/tests-pktbuf/tests-pktbuf.c 196) exp "nvxuO*6o3C=a6g7]" was "nvxu"
.
ng_pktbuf_tests.test_pktbuf_add__memfull (/home/martine/Repositories/RIOT-OS/RIOT/tests/unittests/tests-pktbuf/tests-pktbuf.c 212) ng_pktbuf_add(NULL, NULL, NG_PKTBUF_SIZE + 1, NG_NETTYPE_UNDEF) was not null.
.
ng_pktbuf_tests.test_pktbuf_add__memfull2 (/home/martine/Repositories/RIOT-OS/RIOT/tests/unittests/tests-pktbuf/tests-pktbuf.c 219) ng_pktbuf_add(NULL, NULL, NG_PKTBUF_SIZE - sizeof(ng_pktsnip_t), NG_NETTYPE_UNDEF) was not null.
.
ng_pktbuf_tests.test_pktbuf_add__memfull3 (/home/martine/Repositories/RIOT-OS/RIOT/tests/unittests/tests-pktbuf/tests-pktbuf.c 230) ng_pktbuf_add(NULL, NULL, (NG_PKTBUF_SIZE / 10) + 4, NG_NETTYPE_UNDEF) was not null.
...
ng_pktbuf_tests.test_pktbuf_realloc_data__pkt_NULL (/home/martine/Repositories/RIOT-OS/RIOT/tests/unittests/tests-pktbuf/tests-pktbuf.c 278) exp 2 was -22
.
ng_pktbuf_tests.test_pktbuf_realloc_data__pkt_wrong (/home/martine/Repositories/RIOT-OS/RIOT/tests/unittests/tests-pktbuf/tests-pktbuf.c 286) exp 2 was -22
.
ng_pktbuf_tests.test_pktbuf_realloc_data__pkt_data_wrong (/home/martine/Repositories/RIOT-OS/RIOT/tests/unittests/tests-pktbuf/tests-pktbuf.c 296) exp 2 was -22
.
ng_pktbuf_tests.test_pktbuf_realloc_data__pkt_users_gt_1 (/home/martine/Repositories/RIOT-OS/RIOT/tests/unittests/tests-pktbuf/tests-pktbuf.c 307) exp 22 was -22
.
ng_pktbuf_tests.test_pktbuf_realloc_data__pkt_next_neq_NULL (/home/martine/Repositories/RIOT-OS/RIOT/tests/unittests/tests-pktbuf/tests-pktbuf.c 318) exp 22 was -22
.
ng_pktbuf_tests.test_pktbuf_realloc_data__size_0 (/home/martine/Repositories/RIOT-OS/RIOT/tests/unittests/tests-pktbuf/tests-pktbuf.c 327) exp 12 was -22
.
ng_pktbuf_tests.test_pktbuf_realloc_data__memfull (/home/martine/Repositories/RIOT-OS/RIOT/tests/unittests/tests-pktbuf/tests-pktbuf.c 337) exp 12 was 0
.
ng_pktbuf_tests.test_pktbuf_realloc_data__memfull2 (/home/martine/Repositories/RIOT-OS/RIOT/tests/unittests/tests-pktbuf/tests-pktbuf.c 348) exp 12 was 0
.
ng_pktbuf_tests.test_pktbuf_realloc_data__memfull3 (/home/martine/Repositories/RIOT-OS/RIOT/tests/unittests/tests-pktbuf/tests-pktbuf.c 361) exp 12 was 0
..
ng_pktbuf_tests.test_pktbuf_realloc_data__shrink (/home/martine/Repositories/RIOT-OS/RIOT/tests/unittests/tests-pktbuf/tests-pktbuf.c 377) exp_data == pkt->data
.
ng_pktbuf_tests.test_pktbuf_realloc_data__memenough (/home/martine/Repositories/RIOT-OS/RIOT/tests/unittests/tests-pktbuf/tests-pktbuf.c 394) exp_data == pkt->data
...
ng_pktbuf_tests.test_pktbuf_realloc_data__further_down_the_line (/home/martine/Repositories/RIOT-OS/RIOT/tests/unittests/tests-pktbuf/tests-pktbuf.c 476) ((uint8_t *)pkt1->data) == (((uint8_t *)header->data) + 4)
......make[1]: *** [term] Segmentation fault
make: *** [test] Error 1
make: Leaving directory `/home/martine/Repositories/RIOT-OS/RIOT/tests/unittests'
```

**Comment by haukepetersen** [comment]
Ups, not sure what I did wrong yesterday when testing, now they fail here as well... Will look into this.

File: sys/Makefile (hunk 6 above)
Comment (haukepetersen) [inline sys/Makefile]: no system I used... just hacked it in yesterday...

**Comment by haukepetersen** [comment]
About cleaning this up: I did not expect this PR being merged in this state anyway, I just wanted to PR it, so others might gain something from using it while debugging. But if the general consensus is, that we want to merge it, I clean it up and make it ready for review!

File: sys/net/crosslayer/ng_pktbuf_dynamic/ng_pktbuf_dynamic.c (hunk 5 above)
Comment (haukepetersen) [inline sys/net/crosslayer/ng_pktbuf_dynamic/ng_pktbuf_dynamic.c]: Nope, But I dont see the point why I should add (a lot of) complexity to the packetbuffer implementation just for some unit-tests...

File: sys/net/crosslayer/ng_pktbuf_dynamic/ng_pktbuf_dynamic.c (hunk 4 above)
Comment (haukepetersen) [inline sys/net/crosslayer/ng_pktbuf_dynamic/ng_pktbuf_dynamic.c]: me neither, just took this short-cut to get back to testing...

File: sys/net/crosslayer/ng_pktbuf_dynamic/ng_pktbuf_dynamic.c (hunk 5 above)
Comment (authmillenon) [inline sys/net/crosslayer/ng_pktbuf_dynamic/ng_pktbuf_dynamic.c]: I think the unittests rely on that…

**Comment by haukepetersen** [comment]
> add seperate unittests for pktbuf_dynamic

By the way: what is this supposed to do? I mean the trick about one interface and several implementations is, that identical unittests are applicable!

File: sys/net/crosslayer/ng_pktbuf_dynamic/ng_pktbuf_dynamic.c (hunk 4 above)
Comment (haukepetersen) [inline sys/net/crosslayer/ng_pktbuf_dynamic/ng_pktbuf_dynamic.c]: Just tested (involuntary on some other place): free is ok with NULL ptrs...

**Comment by authmillenon** [comment]
> By the way: what is this supposed to do? I mean the trick about one interface and several implementations is, that identical unittests are applicable!

Same interface is nice, but the different implementations should be tested on the CI system regardless ;-)

**Comment by haukepetersen** [comment]
Ah, now I understand :-)

File: sys/net/crosslayer/ng_pktbuf_dynamic/ng_pktbuf_dynamic.c (hunk 3 above)
Comment (OlegHahm) [inline sys/net/crosslayer/ng_pktbuf_dynamic/ng_pktbuf_dynamic.c]: >  I tried to make this implementation as error prone as possible...

I thought we were all doing this every time we code for RIOT. *scnr*

File: sys/net/crosslayer/ng_pktbuf_dynamic/ng_pktbuf_dynamic.c (hunk 4 above)
Comment (OlegHahm) [inline sys/net/crosslayer/ng_pktbuf_dynamic/ng_pktbuf_dynamic.c]: Manpage says:
> If ptr is NULL, no operation is performed.

**Comment by haukepetersen** [comment]
just rebased, still need to fix the comments above

**Comment by authmillenon** [comment]
Needs adaption to changes introduced in #3458.

**Comment by haukepetersen** [comment]
yes, will adapt it as soon as I find a minute

**Comment by authmillenon** [comment]
Ping?

**Comment by haukepetersen** [comment]
havn't found a minute. And I have to confess, this PR is not very high up my prio list..

**Comment by haukepetersen** [comment]
not planning to work on this anytime soon -> so I marked this as memo and close it for now.

**Pull request closed without merging**